R/git.R

```diff
@@ -101,12 +101,55 @@ git_branches_no_merged <- function(root = NULL, include_master = FALSE) {
                       root = root, check = TRUE)$output
     branches <- c(master, branches)
   }
-  branches <- read.table(text = branches, stringsAsFactors = FALSE, sep = ",",
-                         col.names = c("name", "last_commit"))
+  branches <- utils::read.table(text = branches, stringsAsFactors = FALSE,
+                                sep = ",", col.names = c("name", "last_commit"))
   branches <- branches[branches$name != "gh-pages", ]
-  branches$last_commit_age <-
-    rep(as.integer(Sys.time()), nrow(branches)) - branches$last_commit
-  times <- as.POSIXct(branches$last_commit, origin = "1970-01-01", tz = "UTC")
-  branches$last_commit <- strftime(times)
+  branches$last_commit_age <- calculate_age(branches$last_commit)
+  branches$last_commit <- convert_unix_to_iso_time(branches$last_commit)
   branches
 }
+
+## This gets last 25 commits from master
+## if not master then gets the unmerged commits (limit 25)
+git_commits <- function(branch, root = NULL) {
+  if (branch == "master") {
+    args <- c("log", "--pretty='%h,%cd'", "--date=unix", "--max-count=25",
+              sprintf("refs/remotes/origin/%s", branch))
+  } else {
+    remote_branch <- sprintf("refs/remotes/origin/%s", branch)
+    args <- c("log", "--pretty='%h,%cd'", "--date=unix", "--max-count=25",
+              sprintf("--cherry refs/remotes/origin/master..%s", remote_branch),
+              remote_branch)
+  }
+  commits <- git_run(args, root = root, check = TRUE)$output
+  commits <- utils::read.table(text = commits, stringsAsFactors = FALSE,
+                               sep = ",", col.names = c("id", "date_time"))
+  commits$age <- calculate_age(commits$date_time)
+  commits$date_time <- convert_unix_to_iso_time(commits$date_time)
+  ## ID can be parsed as an integer by read.table if by chance the id contains
+  ## only numbers
+  commits$id <- as.character(commits$id)
+  commits
+}
+
+calculate_age <- function(times) {
+  rep(as.integer(Sys.time()), length(times)) - times
+}
+
+convert_unix_to_iso_time <- function(times) {
+  strftime(as.POSIXct(times, origin = "1970-01-01", tz = "UTC"))
+}
+
+get_reports <- function(branch, commit, root) {
+  if (branch == "master") {
+    ## Get all reports in commit if on master branch
+    args <- c("ls-tree", "--name-only", "-d", sprintf("%s:src/", commit))
+  } else {
+    ## Get only files which have changed from master copy
+    ## Note this could inclue files as well as directories.
+    ## How can we exclude them?
```

Comment:
Also wanted to discuss this point. 

So I tried checking those which are directories using R but that can't be done if it doesn't exist locally (which can definitely happen here). I can't see an obvious flag to `diff-tree` like there is on `ls-tree` for only returning directories. And similarly piping to `du` to only get dir doesn't work for same reasons as checking for dirs in R doesn't work.

I could do something like get return from `diff-tree` return all changed files and then pipe it to `dirname` to only give the dirname part we are interested in.

Reply:
Git only stores files (not directories) which means if you pipe it to `dirname()` then get the *first* element below `src` you your full list of directories that you are interested in - as you suggest above.

Reply:
Thanks, fixed this in R side now